### ISSUE TITLE
Show Existing Scheduled Posts in Calendar / Datepicker

### ISSUE THREAD

**Issue by supernovia** [WordPress/gutenberg#13713]
When choosing a publish date, it's helpful to know which dates already have posts scheduled.

Calypso has a handy feature in its calendar that circles already-scheduled dates, as well as post counts on mouseover. Users love it and regularly request this feature in WordPress.com's Gutenberg implementation [here](https://github.com/Automattic/wp-calypso/issues/28567). You can see it in action [here](http://cld.wthms.co/A26y6A) (HT: @ashercantrell) 

It might be nice to have this or a similar feature built into Core.

**Comment by sarahmonster** [comment]
This seems like it would be a great idea! Let's see what that might look like in practise:

![schedule a](https://user-images.githubusercontent.com/376315/58669612-b59cc900-8334-11e9-85b4-9afee5d5b1fa.png)
![schedule b](https://user-images.githubusercontent.com/376315/58669613-b59cc900-8334-11e9-92b8-bf6d62392aec.png)
![schedule c](https://user-images.githubusercontent.com/376315/58669614-b6355f80-8334-11e9-93ca-ef3d38070698.png)
![schedule d](https://user-images.githubusercontent.com/376315/58669615-b6355f80-8334-11e9-8da2-213ac2bd1ad9.png)
![schedule e](https://user-images.githubusercontent.com/376315/58669617-b6355f80-8334-11e9-92ef-07fb5088718d.png)


I'm inclined to lean toward a pale blue indicator here. The grey has better contrast, but I'm a bit concerned about the potential for users to confuse this with a disabled state. Whilst this is _kind of_ part of what we're implying here, it's not the whole story, and I don't want users to think they can't select these dates.

If we went with blue as an indicator, we'd need to be super careful the differentiation is sufficient between focussed, selected, and scheduled-post indicators, so it's super obvious what's going on at all times here. (See also: #15929)

Using a square shape or a border would help to ensure colour isn't the sole indicator of meaning here, at the expense of a wee bit of visual polish, but that feels like a fair trade-off.

**Comment by melchoyce** [comment]
> Using a square shape or a border would help to ensure colour isn't the sole indicator of meaning here, at the expense of a wee bit of visual polish, but that feels like a fair trade-off.

Maybe round the corners a little bit to match the above arrow buttons?

**Comment by sdixon194** [comment]
Just a heads up, had a user request this in 2127509-zen. The classic editor does this (with the circles) and they miss the feature.

**Comment by thehenrybyrd** [comment]
Had another request for this in a personal communication with someone who blogs regularly:

> it doesn’t show the dates that other posts are scheduled like calypso does. Very frustrating for someone who writes in advance

Would love to see this happen! :)

**Comment by sarahmonster** [comment]
Here's an alternate option using slightly-rounded square corners:
![schedule f](https://user-images.githubusercontent.com/376315/61048586-262a1180-a3da-11e9-9ba1-2defc024268d.png)

And another option using a more active greyed-out styling that I think may be a better solution here. It is a bit noisier, but it does ensure that the styling has sufficient contrast against the white background, and it doesn't conflict with focus styling:
![schedule g](https://user-images.githubusercontent.com/376315/61048587-26c2a800-a3da-11e9-88f7-5a5e8126dce4.png)

My current inclination is to go with option G for accessibility reasons, but it may be easier and more effective to test these variants in code. @brentswisher is this something you might be able to help with as part of the larger publishing flow changes?

**Comment by brentswisher** [comment]
@sarahmonster Sure, I could take a look at this, we can probably work on this in parallel with the publishing changes as it is only loosely related. How would you like to handle, can you start a branch and add the CSS, and once that is done I can work on pulling the data and applying the styles correctly to the proper dates?

A side note, I would think accessibility-wise iterations A, E or F would be preferred, as the shape is different and it is not relying entirely on color. I am thinking we will have to add some additional aria attributes announcing the information to screen readers. I have some idea the proper way to do that but will need to research a bit to verify I am correct.

**Comment by supernovia** [comment]
Just swinging back to this. Users were used to a circle around a scheduled date -- what would it look like if we did that but in the blue?

**Comment by supernovia** [comment]
👋 This came up again today.  I don't see a huge demand for it, but it'd be a nice little way to make our UX better for folks who love to blog, or who write for businesses.
https://en.forums.wordpress.com/topic/scheduling-calendar/

@melchoyce any thoughts on just running with the circle version (A) like folks had used before? I agree with @brentswisher  -- the different shape makes it clearer.  And if we choose to circle the date instead of using a shaded background, we don't have to worry as much about getting proper contrast with all the different color schemes out there.

**Comment by melchoyce** [comment]
Good points, let's try out A.

**Comment by ChrissiePollock** [comment]
This was brought up again in #2583166-zen.

**Comment by supernovia** [comment]
Noting another request here:
https://twitter.com/flameeyes/status/1243605582898237444

Do we need to do anything else to help this along?

**Comment by marekhrabe** [comment]
This is a popular feature from the editor on WordPress.com. I count myself towards fans and people who miss it so I'm looking into doability of this task. 

Source code of WordPress.com version: https://github.com/Automattic/wp-calypso/tree/master/client/components/date-picker
Gutenberg Date Time picker: https://github.com/WordPress/gutenberg/tree/master/packages/components/src/date-time

I like the general API of having `events`. I think that could be a good first step - to introduce a similar support for `events` in Gutenberg.

Separate concern could be how to fill `events` with appropriate data about scheduled posts. There are some tricky bits I can think of right away:
- Which post types do we list? You can schedule other things than posts. Probably good to show the same post type as the edited post? Or always posts?
- User permissions - not all users can see all posts.
- Pagination - fetching ALL scheduled posts at once is naive and we probably want to base it on user flipping calendar months which requires adding more callbacks to the calendar component. Possibly a loading state to know they are looking at month with still unknown scheduled posts?
- Component location. In believe we can't add WP-tied logic into `@wordpress/components`. I think we need some wrapper or HOC like `withScheduledPosts` placed in a different package to respect the separation of concerns.

**Comment by marekhrabe** [comment]
This is what shows up when I use the highlight feature of the `react-dates` library without any additional styling in Gutenberg. I just set it to highlight every fifth day, nothing more clever yet.

<img width="272" alt="Screenshot 2020-04-20 at 16 39 43" src="https://user-images.githubusercontent.com/156676/79993277-e0e05780-84b4-11ea-8f5c-ac392c2bb3cf.png">

**Comment by marekhrabe** [comment]
I'm starting with small steps: the first is adding the `DateTimePicker` into the Storybook so we can test the component isolated for a quicker development. https://github.com/WordPress/gutenberg/pull/22024

Next up: I'll try to extend the component to accept an array of events, probably similar to the Calypso implementation.

**Comment by marekhrabe** [comment]
Not connected to any API just yet — #22032 adds the ability for highlighting days in the calendar.

![2020-05-04 18 18 54](https://user-images.githubusercontent.com/156676/80989586-6c52d480-8e35-11ea-8577-f6ee76e45b79.gif)

**Comment by marekhrabe** [comment]
#### Needs review:

- [ ] #22024 Storybook: Add DateTimePicker
- [ ] #22032 DateTimePicker: Add support for highlighting days

#### Next steps:

After those two are in, we can start thinking about connecting the highlighting feature to the API with scheduled posts.

**Comment by earnjam** [comment]
Something to consider here, this becomes mostly useless and a bit of a nuisance if you have a high-volume site with lots of scheduled posts (i.e. every day or almost every day). In those cases it would be better suited to offload this functionality to a more robust publishing plugin like Edit Flow. 

I think it will be important for the more enterprise-level publishing sites to have some method to be able to disable this feature.

**Comment by dwainm** [comment]
In the case of very busy sites this should be dismissible via a filter. It is useful for bloggers and those with a reasonable amount of scheduled posts.

**Comment by supernovia** [comment]
Happy to see some discussion and work on this. Thank you @marekhrabe .

**Comment by supernovia** [comment]
Just noting another instance / request:
https://wordpress.com/forums/topic/calendar-changes/

**Comment by kcswong** [comment]
Had a request for this in chat, mentioned that it's vital for their site: 6791150-hc

**Comment by supernovia** [comment]
Checking in on this. We've got several requests here on our end:
https://github.com/Automattic/wp-calypso/issues/42986

**Comment by supernovia** [comment]
Another request here. https://wordpress.com/forums/topic/new-editor-new-interface/#post-3554770

Thank you @mtias :)

**Comment by aleone89** [comment]
More request here: hc-14523525 and zen-3326301

**Comment by retrofox** [comment]
> This seems like it would be a great idea! Let's see what that might look like in practise:
> 
> ![schedule a](https://user-images.githubusercontent.com/376315/58669612-b59cc900-8334-11e9-85b4-9afee5d5b1fa.png)

Hi @sarahmonster, could you provide more details about the circle border color of the `A` option, according to [this comment](https://github.com/WordPress/gutenberg/issues/13713#issuecomment-566750959)?

**Comment by supernovia** [comment]
Thanks @retrofox. I'm not sure whether Sarah is working on this still. Do you need help from someone else?

We had another request over Twitter just now.
https://twitter.com/MedievalLitura/status/1308082625156804623

**Comment by retrofox** [comment]
We've created this PR.
https://github.com/WordPress/gutenberg/pull/25428
Feedback is over welcomed :-)

**Comment by supernovia** [comment]
Noting another request here:
https://wordpress.com/forums/topic/calendar-markers/

I'm not sure what to do with the conflicts and what not, but if there's something I can do let me know!

**Comment by simison** [comment]
Needs https://github.com/WordPress/gutenberg/issues/21820 and then https://github.com/WordPress/gutenberg/pull/25428 — both stalled at the moment.
